### PR TITLE
py: Rework and compress bytecode prelude

### PR DESCRIPTION
**Pull request by dpgeorge**
Background: the format of a bytecode function has evolved (since the start of this project) on an as-needed basis.  So the current format is a bit messy and in some cases suboptimal.  This PR aims to clean up the bytecode function format and compress it a little.

The original motivation for this was because the first 6 bytes of the bytecode header are usually mostly 0, which is wasteful.  In my attempts to improve this I ended up reworking the whole header, so that's why this PR is rather large.

-------------

Original format of bytecode (header plus opcodes; see py/bc.h):
```
//  n_state         : var uint
//  n_exc_stack     : var uint
//  scope_flags     : byte
//  n_pos_args      : byte          number of arguments this function takes
//  n_kwonly_args   : byte          number of keyword-only arguments
//  n_def_pos_args  : byte          number of default positional arguments
//
//  code_info_size  : var uint |    code_info_size counts bytes in this chunk
//  simple_name     : var qstr |
//  source_file     : var qstr |
//  <line number info>         |
//  <word alignment padding>   |    only needed if bytecode contains pointers
//
//  local_num0      : byte     |
//  ...             : byte     |
//  local_numN      : byte     |    N = num_cells
//  255             : byte     |    end of list sentinel
//  <bytecode>                 |
```
It's awkward to decode this and some entries are rarely used (eg not many functions have keyword-only args, or closed over cells).  As pointed out in #4378, `code_info_size` includes the size of itself which makes it hard to compute and encode (also complicates decoding).  Furthermore, when calling a bytecode function the `n_state` is decoded 3 times (once to work out how much memory to allocate for the function state, once to parse the arguments, and once at the start of the VM to compute the pointer to the local variables).

------------

New format (all done by this PR):
```
//  fun_signature   : var uint
//      contains six values interleaved bit-wise as: xSSSSEAA [xFSSKAED repeated]
//          x = another byte follows
//          S = n_state - 1         number of entries in Python value stack
//          E = n_exc_stack         number of entries in exception stack
//          F = scope_flags         four bits of flags, MP_SCOPE_FLAG_xxx
//          A = n_pos_args          number of arguments this function takes
//          K = n_kwonly_args       number of keyword-only arguments
//          D = n_def_pos_args      number of default positional arguments
//
//  init_size       : var uint
//      contains two values interleaved bit-wise as: xIIIIIIC repeated
//          x = another byte follows
//          C = n_cells             number of bytes/cells in closure section
//          I = n_info              number of bytes in source info section
//
//  closure section:
//      local_num0  : byte
//      ...         : byte
//      local_numN  : byte          N = n_cells-1
//
//  source info section:
//      simple_name : var qstr
//      source_file : var qstr
//      <line number info>
//
//  <word alignment padding>        only needed if bytecode contains pointers
//
//  <bytecode>
```
The header now starts with 2 variably encoded uints, `fun_signature` and `init_size`.

`fun_signature` contains 6 numbers packed into it via bit-wise interleaving.  The interleaving is chosen based on an entropy analysis of around 150,000 bytecode functions from the CPython standard library.  Optimal Shannon coding would need about 7.1 bits on average to encode these 6 numbers.  The scheme I came up with requires about 10.5 bits on average, which is pretty good given the relative simplicity of the compression technique.

In the end what this means is that most functions which originally took 6 bytes to encode these 6 numbers now need only 1 byte (in 80% of cases).

The `init_size` number encodes 2 items, which are the byte-size of the following 2 sections in the header, the "closure section" and the "source info section".  After decoding this variable uint it's possible to skip over one or both of these sections very easily.  This scheme saves about 2 bytes on average compared to the original format.

Bytecode follows these 2 sections (optionally aligned to machine word if persistent code is disabled).

**In summary: the new format saves about 7 bytes on average for each bytecode function.**  This is pretty significant.  Eg the esp32 port has about 120 frozen bytecode functions, so that saves around 840 bytes of flash.  More importantly this new format saves RAM when functions are compiled to RAM.

### CODE REVIEW

**Comment by dpgeorge** [comment]
Code size change for this PR is:
```
   bare-arm:  +400 +0.607% 
minimal x86:   -20 -0.013% [incl +4(data)]
   unix x64:  +552 +0.111% [incl -224(data)]
unix nanbox:  +140 +0.032% [incl -176(data)]
      stm32:  -196 -0.053% PYBV10
     cc3200:  +416 +0.225% 
    esp8266:  +360 +0.055% 
      esp32:  -680 -0.061% GENERIC[incl -1072(data)]
        nrf:  +208 +0.143% pca10040
       samd:  +388 +0.382% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
So... it's quite a large **increase** for some ports, notably those that don't have any frozen bytecode.  This is because of the increased code size required to encode/decode the new format (mainly the way the bit-wise interleaving works).

For ports that have substantial frozen code (stm32, esp32) the net change is a decrease, because the savings in the frozen bytecode more than make up for the increase in the code size.

As for performance, it seems that it is slightly **improved**.  On PYBv1.0 I measure:
```
diff of scores (higher is better)
N=100 M=100               perf-a    ->  perf-b            diff      diff% (error%)
bm_chaos.py                  292.98 ->     296.66 :      +3.68 =  +1.256% (+/-0.01%)
bm_fannkuch.py                72.47 ->      72.83 :      +0.36 =  +0.497% (+/-0.01%)
bm_float.py                 4726.28 ->    4799.18 :     +72.90 =  +1.542% (+/-0.00%)
bm_hexiom.py                  34.46 ->      35.70 :      +1.24 =  +3.598% (+/-0.00%)
bm_nqueens.py               4137.91 ->    4206.91 :     +69.00 =  +1.668% (+/-0.00%)
bm_pidigits.py               630.62 ->     638.39 :      +7.77 =  +1.232% (+/-0.14%)
builtin_sum_list.py          955.20 ->     960.75 :      +5.55 =  +0.581% (+/-0.00%)
builtin_sum_range.py         969.36 ->     975.00 :      +5.64 =  +0.582% (+/-0.00%)
misc_aes.py                  340.70 ->     359.33 :     +18.63 =  +5.468% (+/-0.01%)
misc_mandel.py              2979.00 ->    2971.91 :      -7.09 =  -0.238% (+/-0.00%)
misc_pystone.py             1839.18 ->    1910.56 :     +71.38 =  +3.881% (+/-0.00%)
misc_raytrace.py             295.60 ->     303.43 :      +7.83 =  +2.649% (+/-0.00%)
viper_call0.py               576.82 ->     576.85 :      +0.03 =  +0.005% (+/-0.00%)
viper_call1a.py              550.34 ->     550.39 :      +0.05 =  +0.009% (+/-0.01%)
viper_call1b.py              434.92 ->     434.93 :      +0.01 =  +0.002% (+/-0.00%)
viper_call1c.py              439.48 ->     439.48 :      +0.00 =  +0.000% (+/-0.00%)
viper_call2a.py              536.30 ->     536.29 :      -0.01 =  -0.002% (+/-0.01%)
viper_call2b.py              377.28 ->     377.29 :      +0.01 =  +0.003% (+/-0.00%)
```
The improvement is probably because the bytecode header is now faster to decode (eg `n_state` is decode only once now, compared to three times before).  Also this decoding is inlined in the code rather than calling out to `mp_decode_uint()` multiple times.

On PYBD-SF2 @ 120MHz I measure:
```
diff of scores (higher is better)
N=500 M=100                  pybd-a ->     pybd-b         diff      diff% (error%)
bm_chaos.py                  300.78 ->     314.53 :     +13.75 =  +4.571% (+/-0.05%)
bm_fannkuch.py                10.29 ->      10.92 :      +0.63 =  +6.122% (+/-0.05%)
bm_float.py                 4609.20 ->    4812.33 :    +203.13 =  +4.407% (+/-0.04%)
bm_hexiom.py                  37.76 ->      40.15 :      +2.39 =  +6.329% (+/-0.02%)
bm_nqueens.py               3825.37 ->    4223.26 :    +397.89 = +10.401% (+/-0.03%)
bm_pidigits.py               713.80 ->     771.16 :     +57.36 =  +8.036% (+/-0.07%)
builtin_sum_list.py         1050.43 ->    1050.30 :      -0.13 =  -0.012% (+/-0.39%)
builtin_sum_range.py         873.91 ->    1010.30 :    +136.39 = +15.607% (+/-0.46%)
misc_aes.py                  372.56 ->     400.89 :     +28.33 =  +7.604% (+/-0.07%)
misc_mandel.py              3168.75 ->    3309.17 :    +140.42 =  +4.431% (+/-0.05%)
misc_pystone.py             1923.01 ->    1996.50 :     +73.49 =  +3.822% (+/-0.04%)
misc_raytrace.py             301.66 ->     313.27 :     +11.61 =  +3.849% (+/-0.05%)
viper_call0.py               562.71 ->     603.41 :     +40.70 =  +7.233% (+/-1.33%)
viper_call1a.py              605.76 ->     607.91 :      +2.15 =  +0.355% (+/-0.61%)
viper_call1b.py              425.32 ->     418.03 :      -7.29 =  -1.714% (+/-0.43%)
viper_call1c.py              430.10 ->     446.23 :     +16.13 =  +3.750% (+/-0.88%)
viper_call2a.py              599.04 ->     607.72 :      +8.68 =  +1.449% (+/-0.18%)
viper_call2b.py              373.70 ->     384.57 :     +10.87 =  +2.909% (+/-0.48%)
```

So, even though code size is increased, I think this patch is worth it.  It decreases RAM usage quite significantly, decreases size of .mpy files, and improves performance by a few percent.

**Comment by dpgeorge** [comment]
> decreases size of .mpy files

For example, the `lcd160cr.py` file was originally 7487 bytes, and with this PR is down to 7114 bytes, a decrease of about 5%.

**Comment by nevercast** [comment]
Great PR Damien. I'm surprised that this takes less time to decode with all the bitwise work required, but as you stated, other code paths are more optimal now. I honestly expected the decoding to be a little worse.

The project I spend most my time on is quite large, is defined in a functional approach (this lots of small single use functions), and it's all frozen. So I'd be interested in testing this on my application. 

Is the ram usage saving just when code is not frozen, or is it saving space due to lookups somewhere, even for functions that are in flash ?

**Comment by dpgeorge** [comment]
> I'm surprised that this takes less time to decode with all the bitwise work required,

If we compare old with new, in the old case the first part of the prelude was decoded as (to get 6 numbers):
- decode uint (this requires a call to `mp_decode_uint()` which has a loop to read in the uint, usually just 1 byte but could be more)
- decode another uint
- load 4 bytes

With this PR the same procedure is now, in the common case (80% of the time):
- read 1 byte
- 3 numbers are extracted from this 1 byte using bit shifting
- remaining 3 numbers are 0

Similarly the remaining parts of the prelude are, in the common case, simpler to decode now.

> Is the ram usage saving just when code is not frozen, or is it saving space due to lookups somewhere, even for functions that are frozen?

About 6-7 bytes are saved per function, for all bytecode functions.  This means 1) functions in RAM; 2) functions frozen in ROM; 3) functions in .mpy files.

**Comment by nevercast** [comment]
Just so we are clear, if a user updates MicroPython to a point after when this commit lands, and they have frozen mpy files in their flash, those files will fail to load?

Just wanting to be clear about what things will break with this internal change.

I'm in half a mind to say "This should be MicroPython 2.0 milestone" and yet I really don't want to wait that long to start seeing the benefits of this PR.

What's your thoughts on breaking changes in this case @dpgeorge ?

**Comment by mikewadsten** [comment]
@nevercast If someone updates to this code once it lands, then they're also getting this commit, which will immediately "invalidate" their existing bytecode: https://github.com/micropython/micropython/commit/5716c5cf65e9b2cb46c2906f40302401bdd27517

And along the same lines as your own comment here (https://github.com/micropython/micropython/issues/5140#issuecomment-535324021), my experience is that MicroPython makes no guarantees of backward compatibility in either bytecode or C source code, either between tagged releases or even in the middle of a single point release (or MPY format version value).

**Comment by nevercast** [comment]
Ah, I didn't see the MPY version bump! Thanks.

**Comment by peterhinch** [comment]
> Just so we are clear, if a user updates MicroPython to a point after when this commit lands, and they have frozen mpy files in their flash, those files will fail to load?

To expand on @mikewadsten's comments I don't think this can happen. If they have frozen mpy files in flash it's because they've compiled a build. Such a user, wanting to update firmware, will pull updated source and recompile. If you install a binary your frozen modules are replaced by those in the binary.

**Comment by nevercast** [comment]
> If they have frozen mpy files in flash it's because they've compiled a build

Not frozen MPY files, just mpy files on a file system.

**Comment by dpgeorge** [comment]
> MicroPython makes no guarantees of backward compatibility in either bytecode or C source code, either between tagged releases or even in the middle of a single point release

That's correct.  We try very hard to have backwards compatibility for Python code (ie the Python-level API), but also doing this for C-level API would make it really hard to move forward (because MicroPython, by its nature, exposes a lot of internal structs and details like that which inevitably change).

The MPY version (affecting precompiled .mpy files) will change at most once between tagged releases.  Yes this technically (in the semver sense) would force a major version increase of MicroPython, but, again, that would put too many restrictions on development of the code base.  So the MPY version/format is considered an "internal detail" and users would need to recompile .mpy files after a version update.  The MPY version is really there just to prevent old .mpy files accidentally running on newer firmware (and crashing randomly).

File: tools/mpy-tool.py [hunk 1]
```diff
@@ -152,21 +152,58 @@ def decode_uint(bytecode, ip):
...
+        z = read_byte()
+        # xFSSKAED
+        n_state |= (z & 0x30) << (2 * n);                       \
+        n_exc_stack |= (z & 0x02) << n;                         \
+        scope_flags |= ((z & 0x40) >> 6) << n;                  \
+        n_pos_args |= (z & 0x4) << n;                           \
+        n_kwonly_args |= ((z & 0x08) >> 3) << n;                \
+        n_def_pos_args |= (z & 0x1) << n;                       \
+        n += 1
+    n_state += 0*n_pos_args + 1;                                  \
```
Comment (mikewadsten) [inline tools/mpy-tool.py]: Isn't this just `n_state += 1`?

File: tools/mpy-tool.py [hunk 2]
```diff
@@ -152,21 +152,58 @@ def decode_uint(bytecode, ip):
...
-    ip, n_exc_stack = decode_uint(bytecode, ip)
-    scope_flags = bytecode[ip]; ip += 1
-    n_pos_args = bytecode[ip]; ip += 1
-    n_kwonly_args = bytecode[ip]; ip += 1
-    n_def_pos_args = bytecode[ip]; ip += 1
-    ip2, code_info_size = decode_uint(bytecode, ip)
-    ip += code_info_size
-    while bytecode[ip] != 0xff:
+    def local_read_byte():
+        nonlocal bytecode, ip
```
Comment (mikewadsten) [inline tools/mpy-tool.py]: I don't know if we're intended to keep `tools/mpy-tool.py` Python 2 compatible, but `nonlocal` breaks that if so.

File: tools/mpy-tool.py (hunk 1 above)
Comment (dpgeorge) [reply tools/mpy-tool.py]: Yes indeed.  I originally had it as `n_state += n_pos_args + 1` to improve compression but it was more complex to decode when only `n_state` is needed.  So this should be simplified.

File: tools/mpy-tool.py (hunk 2 above)
Comment (dpgeorge) [reply tools/mpy-tool.py]: Yeah I think it's good to retain Python 2 compatibility for a while.  Thanks for pointing this out.

**Comment by dpgeorge** [comment]
Thanks for the feedback.  This was merged.
